peepingtom/io_/read/star/star.py

```diff
@@ -1,82 +1,18 @@
-    # invert them so that we return matrices which transform reference onto particle
-    if convention['rotate_reference'] is False:
-        rotation_matrices = rotation_matrices.transpose((0, 2, 1))
-    return rotation_matrices
-
-
-def read_star(star_path, data_columns=[], name_regex=None, **kwargs):
-    """
-    read a starfile into one or multiple ParticleBlocks
-    data_columns: a list of column names to include as properties
-    name_regex: a regex pattern to use to guess the name of individual datasets
-    """
-    raw_dfs = starfile.read((star_path), always_dict=True)
-
-    # metadata = {}
-    dataframes = []
-    # check if relion3.1 format
-    if list(raw_dfs.keys()) == ['optics', 'particles']:
-        # TODO: use optics metadata
-        # metadata = raw_dfs['optics']
-        dataframes.append(raw_dfs['particles'])
-    else:
-        # assume every key is a different dataset (older version)
-        # TODO: make more specific
-        dataframes.extend(list(raw_dfs.values()))
-
-    # create particles
-    datablocks = []
-    for dataframe in dataframes:
-        # try every type of starfile convention
-        for star_type, params in star_types.items():
-            # check if the necessary columns exist
-            if all(colname in dataframe.columns for colname in params['coords']):
-                # split and guess names
-                named_dfs = split_and_name(dataframe, split_by=params['split_by'],
-                                           basename=star_path, name_regex=name_regex)
-                for name, df in named_dfs:
-                    # extract the data and convert it to peepingtom format
-                    raw_positions = df[params['coords']]
-                    raw_shifts = df.get(params['shifts'], 0)
-                    raw_eulers = df[params['angles']]
-                    raw_properties = df[data_columns]
-
-                    positions = (raw_positions + raw_shifts).to_numpy()
-                    orientations = euler2matrix(raw_eulers, params['angle_convention'])
-                    properties = {key: df[key].to_numpy() for key in raw_properties}
-
-                    datablocks.append(ParticleBlock(positions, orientations, properties, name=name))
-                # break of out loop if something was found
-                break
-        else:
-            raise ValueError(f'dataframe with columns {list(dataframe.columns)} could not be read')
-
-    return datablocks
+    raw_data = starfile.read(star_path, always_dict=True)
+
+    failed_reader_functions = []
+    for style, reader_function in reader_functions.items():
+        try:
+            particle_blocks = reader_function(raw_data)
+            return particle_blocks
+        except ValueError:
+            failed_reader_functions.append((style, reader_function))
+    raise NotImplementedError(f'Failed to parse {star_path} using {failed_reader_functions}')
```

Comment:
I don't think is necessarily a `NotImplementedError`. The file may be corrupt, for example. Using the more generic `ValueError` allows us to catch this higher up together with the rest of IO.